core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java

```diff
@@ -186,16 +194,31 @@ public void accept(Response<AiMessage> message) {
             toolExecutors = methodCreateInfo.getToolExecutors();
         }
 
-        Response<AiMessage> response = toolSpecifications == null
-                ? context.chatModel.generate(messages)
-                : context.chatModel.generate(messages, toolSpecifications);
-        log.debug("AI response obtained");
+        Response<AiMessage> response;
+
+        if (cache != null) {
+            log.debug("Attempting to obtain AI response from cache");
+
+            Embedding query = context.embeddingModel.embed(userMessage.text()).content();
```

Comment:
IIUC here we should also check that there is no previous conversation (no user message in the memory), because we want to only support caching for the first message, right? Or are you checking this elsewhere already?

Reply:
The cache is configurable with the property `quarkus.langchain4j.cache.max-size` (default 1). Should this be hidden from developers?

Reply:
Maybe we're not understanding each other - cache max size is the maximum number of prompts stored in it, right?

But here I am talking about the fact that we should only attempt to use the cache when it's for the first prompt in the conversation, so that we  avoid the situation that you described in https://github.com/quarkiverse/quarkus-langchain4j/issues/637#issuecomment-2144429198

Reply:
Yes, `max-size` is the maximum number of prompts stored, where prompts should be the combination of `SystemMessage + UserMessage`. Each cache as id, if possible the id of the cache is unique (for example if we are in a `RequestScope` context), if not the id will always be the value `default`.

Id with context:
`<ContextId>#<Class>.<Method>`

Id without context:
`default`

However, I don't think we understand each other (but that's my problem :dizzy_face:). 
To avoid returning a wrong cached response like in the example in [#637 (comment)](https://github.com/quarkiverse/quarkus-langchain4j/issues/637#issuecomment-2144429198), is it not enough to have a `max-size` of 1?

Reply:
> Having said that, I don't think we understand each other (but that's my problem 😵). To avoid returning a wrong cached response like in the example in [#637 (comment)](https://github.com/quarkiverse/quarkus-langchain4j/issues/637#issuecomment-2144429198), is it not enough to have a `max-size` of 1?

A single `Cache` instance contains prompts (system+user message pairs) from MULTIPLE conversations; and from each such conversation, it stores only the initial prompt, do I get that right? Therefore, the `max-size` refers to the number of conversations from which we store the initial prompt? But then that has nothing to do with how many messages have passed in the current conversation no?

Reply:
> A single Cache instance contains prompts (system+user message pairs) 

Correct.

> from MULTIPLE conversations;

Depends on the `id` used to get the `Cache`. It's working like what we have today for the `ChatMemory`.
If your application can have a scope, the `id`  of the `Cache` will be composed by the `unique id of the request + class-name + method-name`, this means that each user will have an instance of a `Cache` for each method in the class annotated with `@RegisterAiService`.

If your application doesn't have a scope, the value id `default` will be used, this means that the `Cache` is only one for all `@RegisterAiService` (all the users has the same `Cache`).

> it stores only the initial prompt

The default implementation is a `MessageWindow` with a `max-size` of `1`, which means that only the most recent prompt is stored.

Reply:
My first approach doesn't work and now I understand your point, so I'll work on improving it.
The idea is to have a cache for each method (if annotated) of an `@RegisterAiService` interface that is shared by all users. The number of prompts stored in the cache is equal to the `max-size` configuration value **(default 1)**.

**Example:**

```java
package org.acme.example;

@RegisterAiService
@CacheResult
@SystemMessage("You are a poet")
public interface LLMService {

	UserMessage("Write a poem of 5 lines about {poem}")
	public String fiveLinesPoem(String poem);
	
	UserMessage("Write a poem of 10 lines about {poem}")
	public String tenLinesPoem(String poem);
}
```

In this case, two caches are created:

- `Cache#1` with id: `#org.acme.exampleLLMService.fiveLinesPoem`
- `Cache#2` with id: `#org.acme.exampleLLMService.tenLinesPoem`

Now suppose we have two users, `ADM` and `JM`, who want to create poems.

`ADM` calls 
 -> fiveLinesPoem("dog"); 
 -> No value in the cache;
 -> Calls the LLM;
 -> Stores the response in the `Cache#1` **[MAX-SIZE REACHED]**;
 

`JM` calls 
 -> tenLinesPoem("dog"); 
 -> No value in the cache;
 -> Calls the LLM;
 -> Stores the response in the `Cache#2` **[MAX-SIZE REACHED]**;


`ADM` calls
 -> tenLinesPoem("dog"); 
 -> Response found in the `Cache#2` from the call executed by `JM`;
 -> Return the cached value;
 
`ADM` calls
 -> tenLinesPoem("cat"); 
 -> No value in the cache;
 -> Calls the LLM;
 -> Value not stored in `Cache#1` for **MAX-SIZE REACHED**! 
 
 
The TTL configuration can be used to clean up the cached values. 
What do you think about it? Do you see any potential problems? If it is okay with you, I can start with these changes.

Reply:
One point I see here is that the `ADM` user can have access to `JM` responses, not so if we want to make the cache user-bound (figure out how to do this).

Reply:
Maybe something like `@MemoryId` could help.

Reply:
This idea of implementation has a limitation, in the example above, the cache will be available only during the request context, this means that if you will do more request to the LLM in the same context you can get a cached response, but when the request will end, the cache will be cleared and everything will start from zero.

The same behavior is present when a developer uses an `@ApplicationScope`, but in this case there is something worse which is that they will be stored new cache for each call until  the applicatikon memory is saturate (and of course this is not acceptable).

A solution could be to have another annotation `@CacheId` that is mandatory if you are in the `@ApplicationScope` scenario, in this case the number of cache will be equal to the users.

This doesn't solve the out-of-memory problem, because if the number of users is too high, the problem persists, so the developer must have a fine control of the cache to avoid this problem (this means to have the same behavior of the ChatMemory).

I don't know if I'm making the implementation too complicated and you have something in mind that solves the problem in a much simpler way, in this case your help is welcome

Reply:
Hmm I thought we were going to make the cache shared between users, not scoped to each user individually. 

Yes, that has potential security implications, but I don't find it very useful if we only consider cached results for each user separately - how often will a single user repeat the same questions again? Not very often, I think.

Scoping the cache to an HTTP request is too limiting too - one HTTP request usually involves just one prompt. We need to keep the cached responses for longer so that they can be reused in later conversations. If you scope it to a HTTP request, you dispose of it immediately after one request.

My understanding was we would have a cache for each AI service method, but its life would be application-scoped and store the first prompt of every conversation that starts by calling that method.

Reply:
> My understanding was we would have a cache for each AI service method, but its life would be application-scoped and store the first prompt of every conversation that starts by calling that method.

Ok, so in this case the id of the cache will always be: `#<ClassName>.<MethodName>`
In the example above: `#org.acme.exampleLLMService.fiveLinesPoem` and `#org.acme.exampleLLMService.tenLinesPoem` for all the users.

> How to avoid memory problems, that's an important question - set a reasonable TTL and periodically clean it, or I'm not sure we could use weak references to have the GC clean up stuff over time.

Yes, in this case it might be better to set a default TTL and of course document the fact that the developer can find an out-of-memory problem if the TTL value is too high.

Reply:
Ok and the default `max-size` will surely be more than one, because we will store prompts from potentially different users. So actually, the `max-size` will probably be the main safeguard against eating too much memory

Reply:
Ok, let me go back to the example with the last changes we have in mind:

```java
package org.acme.example;

@RegisterAiService
@CacheResult
@SystemMessage("You are a poet")
public interface LLMService {

	@UserMessage("Write a poem of 5 lines about {poem}")
	public String fiveLinesPoem(String poem);
	
	@UserMessage("Write a poem of 10 lines about {poem}")
	public String tenLinesPoem(String poem);
}
```

**Configuration:**
```properties 
max-size: 2
ttl: 5m
```

`ADM` and `JM`, wants to create poems.

**Two caches are created:**
```
Cache#1:  #org.acme.exampleLLMService.fiveLinesPoem
Cache#2:  #org.acme.exampleLLMService.tenLinesPoem
```

**Steps:**
1) ADM calls
-> fiveLinesPoem("dog");
-> No value in the cache;
-> Calls the LLM;
-> Stores the response in the `Cache#1`

2) JM calls
-> fiveLinesPoem("cat");
-> No value in the cache;
-> Calls the LLM;
-> Stores the response in the `Cache#1` **[MAX-SIZE REACHED]**

3) ADM calls
-> fiveLinesPoem("turtle");
-> No value in the cache;
-> Calls the LLM;
-> The value is NOT cached [max size reached]

4) ADM calls
-> fiveLinesPoem("cat");
-> Return the value in the cache.

The two cached responses will be deleted after 5 minutes.

Reply:
Yes this makes sense to me. But let's ask @cescoffier to confirm

Reply:
Any suggestion about that? @cescoffier